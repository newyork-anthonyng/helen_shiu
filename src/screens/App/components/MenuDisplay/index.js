import React, { Component, Children } from 'react';
import T from 'prop-types';

class MenuDisplay extends Component {
  componentDidMount() {
    let nextChildIndex = 0;
    if (this.props.shouldFocusOnLastElement) {
      nextChildIndex = Children.count(this.props.children) - 1;
    }

    this.refs[`child-${nextChildIndex}`].focus();
  }

  handleKeyDown = (index) => {
    return (e) => {
      let nextFocusableChildIndex;
      const childrenCount = Children.count(this.props.children);

      switch (e.keyCode) {
        case 38: // UP
          nextFocusableChildIndex = index - 1;
          if (nextFocusableChildIndex < 0) {
            nextFocusableChildIndex = childrenCount - 1;
          }
          break;
        case 40: // DOWN
          nextFocusableChildIndex = index + 1;
          if (nextFocusableChildIndex >= childrenCount) {
            nextFocusableChildIndex = 0;
          }
          break;
      }

      if (nextFocusableChildIndex !== undefined) {
        this.refs[`child-${nextFocusableChildIndex}`].focus();
      }
    }
  }

  renderChildren = () => {
    return Children.map(this.props.children, (child, index) => {
      return React.cloneElement(child, {
        ref: `child-${index}`,
        tabIndex: -1,
        onKeyDown: this.handleKeyDown(index),
      });
    });
  }

  render() {
    return (
      <div>
        {this.renderChildren()}
      </div>
    );
  }
}

MenuDisplay.defaultProps = {
  shouldFocusOnLastElement: false,
};

MenuDisplay.propTypes = {
  shouldFocusOnLastElement: T.bool,
};

export default MenuDisplay;
