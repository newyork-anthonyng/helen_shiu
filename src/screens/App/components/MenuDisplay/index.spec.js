import React from 'react';
import { shallow, mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import MenuDisplay from './';

const defaultProps = {
  shouldFocusOnLastElement: false,
};

it('should render children correctly', () => {
  const wrapper = shallow(
    <MenuDisplay {...defaultProps}>
      <h1>First</h1>
      <h1>Second</h1>
      <h1>Third</h1>
    </MenuDisplay>
  );

  expect(toJSON(wrapper)).toMatchSnapshot();
});

it('should focus on first element by default', () => {
  const wrapper = mount(
    <MenuDisplay {...defaultProps}>
      <h1>First</h1>
      <h1>Second</h1>
      <h1>Third</h1>
    </MenuDisplay>
  );

  const firstElement = wrapper.find('h1').get(0);

  expect(document.activeElement).toEqual(firstElement);;
});

it('should focus on last element when shouldFocusOnLastElement is true', () => {
  const wrapper = mount(
    <MenuDisplay {...defaultProps} shouldFocusOnLastElement>
      <h1>First</h1>
      <h1>Second</h1>
      <h1>Third</h1>
    </MenuDisplay>
  );

  const lastElement = wrapper.find('h1').get(2);

  expect(document.activeElement).toEqual(lastElement);
});

describe('When pressing down arrow', () => {
  it('should focus on next element', () => {
    const wrapper = mount(
      <MenuDisplay {...defaultProps}>
        <h1>First</h1>
        <h1>Second</h1>
        <h1>Third</h1>
      </MenuDisplay>
    );

    const firstElement = wrapper.find('h1').at(0);
    firstElement.simulate('keydown', { keyCode: 40 });

    const secondElement = wrapper.find('h1').get(1);
    expect(document.activeElement).toEqual(secondElement);
  });

  it('should focus on first element, if focus is currently on last element', () => {
    const wrapper = mount(
      <MenuDisplay {...defaultProps}>
        <h1>First</h1>
        <h1>Second</h1>
        <h1>Third</h1>
      </MenuDisplay>
    );

    const lastElement = wrapper.find('h1').at(2);
    lastElement.simulate('keydown', { keyCode: 40 });

    const firstElement = wrapper.find('h1').get(0);
    expect(document.activeElement).toEqual(firstElement);
  });
});

describe('When pressing up arrow', () => {
  it('should focus on previous element', () => {
    const wrapper = mount(
      <MenuDisplay {...defaultProps}>
        <h1>First</h1>
        <h1>Second</h1>
        <h1>Third</h1>
      </MenuDisplay>
    );

    const lastElement = wrapper.find('h1').at(2);
    lastElement.simulate('keydown', { keyCode: 38 });

    const secondElement = wrapper.find('h1').get(1);
    expect(document.activeElement).toEqual(secondElement);
  });

  it('should focus on last element, if focus is currently on first element', () => {
    const wrapper = mount(
      <MenuDisplay {...defaultProps}>
        <h1>First</h1>
        <h1>Second</h1>
        <h1>Third</h1>
      </MenuDisplay>
    );

    const firstElement = wrapper.find('h1').at(0);
    firstElement.simulate('keydown', { keyCode: 38 });

    const lastElement = wrapper.find('h1').get(2);
    expect(document.activeElement).toEqual(lastElement);
  });
});
