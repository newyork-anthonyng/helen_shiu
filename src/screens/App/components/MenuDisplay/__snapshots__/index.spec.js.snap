// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render children correctly 1`] = `
<div>
  <h1
    onKeyDown={[Function]}
    tabIndex={-1}
  >
    First
  </h1>
  <h1
    onKeyDown={[Function]}
    tabIndex={-1}
  >
    Second
  </h1>
  <h1
    onKeyDown={[Function]}
    tabIndex={-1}
  >
    Third
  </h1>
</div>
`;
